src/python/pants/engine/internals/specs_rules.py

```diff
@@ -446,9 +449,62 @@ def __init__(
         )
 
 
+def _handle_ambiguous_result(
+    request: TargetRootsToFieldSetsRequest,
+    result: TargetRootsToFieldSets,
+    field_set_to_default_to: type[FieldSet],
+) -> TargetRootsToFieldSets:
+    assert len(result.targets) > 1
+    field_set_types = [type(field_set) for field_set in result.field_sets]
+
+    # NB: See https://github.com/pantsbuild/pants/pull/15849. We don't want to break clients as
+    # we shift behavior, so we add this temporary hackery here.
+    if (
+        # (We check for 2 targets because 3 would've been ambiguous pre-our-change)
+        len(result.targets) == 2
+        and PexBinaryFieldSet in field_set_types
+        and PythonSourceFieldSet in field_set_types
+    ):
+        if field_set_to_default_to is PexBinaryFieldSet:
+            warn_or_error(
+                "2.14.0dev1",
+                "referring to a `pex_binary` by using the filename specified in `entry_point`",
+                softwrap(
+                    """
+                        In Pants 2.14, a `pex_binary` can no longer be referred to by the filename that
+                        the `entry_point` field uses.
+
+                        This is due to a change in Pants 2.13, which allows you to use the `run` goal
+                        directly on a `python_source` target without requiring a `pex_binary`. As a
+                        consequence the ability to refer to the `pex_binary` via its `entry_point` is
+                        being removed, as otherwise it would be ambiguous which target to use.
+
+                        Note that because of this change you are able to remove any `pex_binary` targets
+                        you have declared just to support the `run` goal
+                        (usually these are developer scripts), as using `run` on the `python_source` will
+                        have the equivalent behavior.
+
+                        To fix this deprecation, you can use the `pex_binary`'s address to refer to
+                        the `pex_binary`, or set the `[GLOBAL].use_deprecated_pex_binary_run_semantics`
```

Comment:
I'm not following how the option impacts what gets chosen?

Reply:
Down below on line 571:
```py
                PexBinaryFieldSet
                if use_deprecated_pex_binary_run_semantics.val
                else PythonSourceFieldSet,
```

which gets used on line 498 for which target to select.

Reply:
Also, I added a test for this!